### PR TITLE
`Development`: Refactor permission checks for exam access in the server code

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
In `ExamAccessService` the permission checks returned an optional failure instead of directly throwing an `EntityNotFound`/`Conflict`/`AccessForbidden`/…-exception.

### Description
This PR refactors this service to follow the access check style for exercises and courses.

### Steps for Testing
Prerequisites:
- 1 Exam with Instructor, Tutor, Student

This PR affects all server endpoints related to exams. Therefore, please check that you can access the various pages/parts for an exam as before.
No functional changes should have been made.

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage
| Unit | Branch | Line |
| --- | --- | --- |
| `ExamAccessService` | 89% | 97% |

### CODE REVIEW

**Review by krusche — CHANGES_REQUESTED** [review]
Thanks a lot for the great work. This makes the code much more readable and easier to understand! 

In a recent commit, I changed how the `ConflictException` works. It is now closer to the `BadRequestAlertException` and needs some parameters such as an error message. Could you please adapt the places in which you use it. You should receive compile errors.

Thanks again!

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java [hunk 1]
```diff
@@ -16,13 +13,18 @@
...
+import de.tum.in.www1.artemis.web.rest.errors.ConflictException;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
 
 /**
  * Service implementation to check exam access.
  */
 @Service
 public class ExamAccessService {
 
+    private static final String ENTITY_NAME = "exam";
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java:26]:
```suggestion
    private static final String ENTITY_NAME = "Exam";
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 2]
```diff
@@ -694,17 +675,14 @@ private void loadExercisesForStudentExam(StudentExam studentExam) {
...
-        }
+        examAccessService.checkCourseAndExamAndStudentExamAccessElseThrow(courseId, examId, studentExamId);
+
         StudentExam studentExam = studentExamRepository.findById(studentExamId).orElseThrow(() -> new EntityNotFoundException("studentExam", studentExamId));
         if (!studentExam.isSubmitted()) {
             return badRequest();
         }
         if (studentExam.getIndividualEndDateWithGracePeriod().isAfter(ZonedDateTime.now())) {
-            return forbidden();
+            throw new AccessForbiddenException("exam", examId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:685]:
```suggestion
            throw new AccessForbiddenException("Exam", examId);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 3]
```diff
@@ -654,17 +638,14 @@ private void loadExercisesForStudentExam(StudentExam studentExam) {
...
-        }
+        examAccessService.checkCourseAndExamAndStudentExamAccessElseThrow(courseId, examId, studentExamId);
+
         StudentExam studentExam = studentExamRepository.findById(studentExamId).orElseThrow(() -> new EntityNotFoundException("studentExam", studentExamId));
         if (studentExam.isSubmitted()) {
             return badRequest();
         }
         if (studentExam.getIndividualEndDateWithGracePeriod().isAfter(ZonedDateTime.now())) {
-            return forbidden();
+            throw new AccessForbiddenException("exam", examId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:648]:
```suggestion
            throw new AccessForbiddenException("Exam", examId);
```

**Review by MerlinMehmed — DISMISSED** [review]
Code changes look good to me. I also tested exam functionality on TS1. I created an exam with two exercise groups and participated as a student. It all worked fine. Exam assessment and deletion is also okay.
However, I found an issue but it was not introduced by this PR as I reproduced it on TS3 as well. As this is not introduced here, I approve the PR. The issue I faced is described below:
As a student who reviews the exam, I cannot properly do this action. See the screenshot: 
Screenshot from 2022-01-22 12-19-14
The first exercise is ok but the second one (the second box including only ] ) is not properly shown.

**Review by MerlinMehmed — APPROVED** [review]
Reapprove after entity name change.

**Pull request merged**